common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -410,12 +503,17 @@ public MergeStatuses finalizeShuffleMerge(FinalizeShuffleMerge msg) throws IOExc
           + "with the current attempt id %s stored in shuffle service for application %s",
           msg.appAttemptId, appShuffleInfo.attemptId, msg.appId));
     }
-    Map<Integer, AppShufflePartitionInfo> shufflePartitions =
-      appShuffleInfo.partitions.remove(msg.shuffleId);
+    Map<Integer, Map<Integer, AppShufflePartitionInfo>> shuffleSequencePartitions =
+        appShuffleInfo.partitions.get(msg.shuffleId);
+    Map<Integer, AppShufflePartitionInfo> shufflePartitions = shuffleSequencePartitions.get(msg.shuffleSequenceId);
+    if (shufflePartitions != INVALID_SHUFFLE_PARTITIONS) {
+      shuffleSequencePartitions.put(msg.shuffleSequenceId, FINALIZED_SHUFFLE_PARTITIONS);
+      appShuffleInfo.partitions.put(msg.shuffleId, shuffleSequencePartitions);
+    }
     MergeStatuses mergeStatuses;
     if (shufflePartitions == null || shufflePartitions.isEmpty()) {
       mergeStatuses =
-        new MergeStatuses(msg.shuffleId, new RoaringBitmap[0], new int[0], new long[0]);
+        new MergeStatuses(msg.shuffleId, msg.shuffleSequenceId, new RoaringBitmap[0], new int[0], new long[0]);
```

Comment:
We'd return the `MergeStatuses` even if it's a finalize request from an older shuffle?

Reply:
Good catch! Even though we are returning empty mergeStatuses, this is still not right, fixed it by throwing RuntimeException.